java/org/apache/catalina/session/DataSourceStore.java

```diff
@@ -626,15 +626,77 @@ public void save(Session session) throws IOException {
-                            InputStream in = new BufferedInputStream(bis, size);
-                            PreparedStatement preparedSaveSql = _conn.prepareStatement(saveSql)) {
-                        preparedSaveSql.setString(1, session.getIdInternal());
-                        preparedSaveSql.setString(2, getName());
-                        preparedSaveSql.setBinaryStream(3, in, size);
-                        preparedSaveSql.setString(4, session.isValid() ? "1" : "0");
-                        preparedSaveSql.setInt(5, session.getMaxInactiveInterval());
-                        preparedSaveSql.setLong(6, session.getLastAccessedTime());
-                        preparedSaveSql.execute();
+                         InputStream in = new BufferedInputStream(bis, size);
+                         PreparedStatement preparedSaveSql = _conn.prepareStatement(saveSql, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_UPDATABLE)) {
+
+                        // Store auto-commit state
+                        boolean autoCommit = _conn.getAutoCommit();
+
+                        try {
+                            if(autoCommit) {
+                                _conn.setAutoCommit(false); // BEGIN TRANSACTION
+                            }
+
+                            preparedSaveSql.setString(1, getName());
+                            preparedSaveSql.setString(2, session.getIdInternal());
+
+                            ResultSet rs = preparedSaveSql.executeQuery();
+
+                            if(rs.next()) {
+                                // Session already exists in the db; update the various fields
+                                rs.updateBinaryStream(sessionDataCol, in, size);
+                                rs.updateString(sessionValidCol, session.isValid() ? "1" : "0");
+                                rs.updateInt(sessionMaxInactiveCol, session.getMaxInactiveInterval());
+                                rs.updateLong(sessionLastAccessedCol, session.getLastAccessedTime());
+
+                                rs.updateRow();
+                            } else {
+                                // Session does not exist. Insert.
+                                rs.moveToInsertRow();
+
+                                rs.updateString(sessionAppCol, getName());
+                                rs.updateString(sessionIdCol, session.getIdInternal());
+                                rs.updateBinaryStream(sessionIdCol, in, size);
+                                rs.updateString(sessionValidCol, session.isValid() ? "1" : "0");
+                                rs.updateInt(sessionMaxInactiveCol, session.getMaxInactiveInterval());
+                                rs.updateLong(sessionLastAccessedCol, session.getLastAccessedTime());
+
+                                rs.updateRow();
+                            }
+
+                            _conn.commit();
+                        } catch (SQLException sqle) {
```

Comment:
Is it possible to put all the handlers in the same block [1]?  e.g.

```
} catch (SQLException | RuntimeException | Error e) {
    // looks like mostly the same block to me
}
```

We don't need to support Java 6

[1] https://docs.oracle.com/javase/7/docs/technotes/guides/language/catch-multiple.html

Reply:
I would be okay merging them together, but we might want to use different error messages. This is one of the things I wanted to get feedback on: how important is it to say "we got an Error" vs "we got an SQLException", etc. when the actual exception and stack trace will likely be logged? I think it's probably okay to merge these exception handlers together but wanted some feedback about the logging.

Reply:
The Stack Trace would be the same, true, but the exception class and message would still provide the details